File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <set>
#include <queue>

using namespace std;

bool is_one_diff(string a, string b) {
    int diff = 0;
    for(int i=0;i<a.length();i++) {
        if (a[i] != b[i]) {
            if(diff > 1) return false;
            else diff++;
        } 
    }
    if (diff == 1) return true;
    else return false;
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    int l = words[0].length();
    set<string> s(words.begin(), words.end());
    if (!s.count(target)) return answer;

    vector<bool> visited(words.size(), false);
    queue<pair<string,int>> q;
    q.push({begin,0});
    while(!q.empty()) {
        auto [cur_str, cur_dist] = q.front();
        q.pop();
        if(cur_str == target) {
            answer = cur_dist;
            break;
        }
        
        for(int i=0;i<words.size();i++) {
            if (!visited[i] && is_one_diff(words[i], cur_str)) {
                q.push({words[i],cur_dist+1});
                visited[i] = true;
            }
        }
    }

    
    return answer;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.swift
import Foundation

func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
    
    var answer = 0
    
    func isOneDiff(_ s1: String, _ s2: String) -> Bool {
        var diff = 0
        
        let arr1 = Array(s1)
        let arr2 = Array(s2)
        
        for i in 0..<arr1.count {
            if arr1[i] != arr2[i] { diff += 1 }
        }
        
        return diff == 1
    }
    
    var visited = Array(repeating: false, count: words.count)
    
    func dfs(_ cur: String, _ cnt: Int) {
        if cur == target {
            answer = cnt
            return
        }
        
        for (idx, word) in words.enumerated() {
            if !visited[idx] && isOneDiff(cur, word) {
                visited[idx] = true
                dfs(word, cnt+1)
                visited[idx] = false
            }
        }
    }
    
    dfs(begin, 0)
    
    return answer
}